### PR TITLE
Add warnings options

### PR DESCRIPTION
**Pull request by Alex-PLACET**
Implement #88

### CODE REVIEW

File: include/sparrow/dynamic_bitset.hpp
Comment (jjerphan) [inline include/sparrow/dynamic_bitset.hpp:581]: For consistency?
```suggestion
              storage_type(this->compute_block_count(n), value ? block_type(~block_type(0)) : block_type(0)),
```

File: test/test_iterator.cpp
Comment (jjerphan) [inline test/test_iterator.cpp:270]: This change looks unrelated to the scope of this PR, but it being part of it is OK to me.

File: test/test_typed_array.cpp
Comment (jjerphan) [inline test/test_typed_array.cpp:183]: What is the reason for using `int32_t`, here and bellow?

File: test/CMakeLists.txt
Comment (jjerphan) [inline test/CMakeLists.txt:92]: Naive question: do you think it would be worth factorizing flags used by  both `gcc` and `clang`?

File: include/sparrow/buffer.hpp
Comment (jjerphan) [inline include/sparrow/buffer.hpp:1]: It seems this file needs reformatting.

**Review by jjerphan — COMMENTED** [review]
LGTM modulo a few comments and questions.

File: include/sparrow/dynamic_bitset.hpp
Comment (Alex-PLACET) [reply include/sparrow/dynamic_bitset.hpp:581]: Fix a warning about a conversion

File: test/CMakeLists.txt
Comment (jjerphan) [reply test/CMakeLists.txt:92]: How about defining a list for the common ones?

Not that this suggestion need not be addressed at all.

File: test/test_iterator.cpp
Comment (Alex-PLACET) [reply test/test_iterator.cpp:270]: You right, this tests was missing

File: include/sparrow/dynamic_bitset.hpp
Comment (jjerphan) [reply include/sparrow/dynamic_bitset.hpp:581]: My remark was unclear: based [on this change on l.525](https://github.com/xtensor-stack/sparrow/pull/90/files#diff-eb6ccf28ffe0d12cb1cfb2b3d4c3871ee6430fd1d558b86880a69e62788f7f33R541), I was wondering if this line also should be modified similarly?

File: test/test_typed_array.cpp
Comment (Alex-PLACET) [reply test/test_typed_array.cpp:183]: bitmap[] argument is a range_difference_t (aka 'long long'), and so the conversion changes the signedness

**Review by jjerphan — APPROVED** [review]
LGTM.

It seems that a few files aren't formatted (this need not be addressed IMO).

File: include/sparrow/mp_utils.hpp
Comment (Klaim) [inline include/sparrow/mp_utils.hpp:195]: Note that it might be simpler here to just remove the name of `list`.
Also note that `predicate` will always be refered to, so it doesnt  need to be `maybe_unused`.

File: test/CMakeLists.txt
Comment (Klaim) [inline test/CMakeLists.txt:57]: I would like that we try all the flags recommended by Jason Turner: https://github.com/cpp-best-practices/cppbestpractices/blob/master/02-Use_the_Tools_Available.md#compilers
and then progressively remove what we dont want.

For MSVC we also need to add `/Zc:__cplusplus` to make sure the `__cplusplus` define has the right value, see: https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170

Finally, shouldnt we have these flags in cmake toolchains instead of here? Doesnt this will affect users? (not sure what's the best practice regarding this)

File: include/sparrow/contracts.hpp
Comment (jjerphan) [reply include/sparrow/contracts.hpp:97]: > BTW why the need for do/while? Why not just write a block?

An explanation is given in [this thread](https://groups.google.com/g/comp.lang.c/c/xGZxls194mI/m/dEIpTKz2okMJ).

File: include/sparrow/dynamic_bitset.hpp
Comment (Alex-PLACET) [reply include/sparrow/dynamic_bitset.hpp:771]: The issue was the comparison between a signed and unsigned values.

File: include/sparrow/dynamic_bitset.hpp
Comment (Alex-PLACET) [reply include/sparrow/dynamic_bitset.hpp:541]: > error: operand of ? changes signedness: 'int' to 'const block_type' (aka 'const unsigned char')

File: test/CMakeLists.txt
Comment (Alex-PLACET) [reply test/CMakeLists.txt:57]: I activated all warnings, just to test, I finaly deactivated a dozen of them. It's maybe easier to maintain than activating some of them and forgot some.
I let you check the code (I commented the flags to explicitly activate the warnings)

File: include/sparrow/data_type.hpp
Comment (Alex-PLACET) [reply include/sparrow/data_type.hpp]: I forgot to activate the warning as error for msvc, it's fixed now.
I deactivate theses warnings as the header is a 3rd party and we don't have any unit test to validate it, so I prefer to not change the code.

File: include/sparrow/contracts.hpp
Comment (Alex-PLACET) [reply include/sparrow/contracts.hpp]: Ok I reverted

File: include/sparrow/array_data.hpp
Comment (Alex-PLACET) [reply include/sparrow/array_data.hpp]: Ok, reverted 👍

File: include/sparrow/dynamic_bitset.hpp
Comment (Alex-PLACET) [reply include/sparrow/dynamic_bitset.hpp]: Fixed 👍

File: include/sparrow/dynamic_bitset.hpp
Comment (Alex-PLACET) [reply include/sparrow/dynamic_bitset.hpp]: Fixed 👍

File: include/sparrow/mp_utils.hpp
Comment (Alex-PLACET) [reply include/sparrow/mp_utils.hpp]: I fixed the issue for list by just removing the name, but on GCC, the predicate parameter is detected as not used

File: test/CMakeLists.txt
Comment (Alex-PLACET) [reply test/CMakeLists.txt:57]: About where to put these flags, I think we should keep them in our test as the user may not have the same requirements in terms of warnings as errors

File: include/sparrow/contracts.hpp
Comment (Klaim) [inline include/sparrow/contracts.hpp:96]: I think the formatting is a bit extreme for these function calls 😂 
Can you deactivate it for the 3 print calls and if you want to indent it in 2 lines it's fine? Right now it's header to read than the original.

File: include/sparrow/contracts.hpp
Comment (Klaim) [inline include/sparrow/contracts.hpp:16]:
```suggestion
// https://github.com/Klaim/nocontracts/tree/0ffc183f7527213e3f0a57b8dae9befa7c0ca02c
// Modifications: renamed macros
```

File: include/sparrow/contracts.hpp
Comment (Klaim) [reply include/sparrow/contracts.hpp]: I see, not sure if it's still undefined behavior though. Maybe it's not because it's not an infinite loop. Ok lets try that.

File: include/sparrow/data_type.hpp
Comment (Klaim) [reply include/sparrow/data_type.hpp]: Nice, could you add a comment about these warnings being specific to that 3rdpary header?

File: include/sparrow/mp_utils.hpp
Comment (Klaim) [reply include/sparrow/mp_utils.hpp]: Looks like a bug? it's used in the return expression...
Maye add a "TODO: remove `[[mayb_unused]]` once gcc is fixed" in the comment you added there.

File: include/sparrow/contracts.hpp
Comment (Klaim) [reply include/sparrow/contracts.hpp:96]: Maybe also deactivate the formatting around the `do/while` block 🥲

Comment (Alex-PLACET) [reply include/sparrow/contracts.hpp:96]: Oups, sorry I didn't view this auto indent, I will add a NOLINT block

File: include/sparrow/contracts.hpp
Comment (Alex-PLACET) [reply include/sparrow/contracts.hpp]: Done

File: include/sparrow/data_type.hpp
Comment (Alex-PLACET) [reply include/sparrow/data_type.hpp]: Done

File: test/CMakeLists.txt
Comment (Klaim) [inline test/CMakeLists.txt:82]: This one is *not* recommended (see https://github.com/cpp-best-practices/cppbestpractices/blob/master/02-Use_the_Tools_Available.md#msvc ).
I suspect removing this will make the rest work better.

**Review by Klaim — APPROVED** [review]
LGTM except the msvc warning flags but it will be easier for me (who has a windows always available) to check these and patch after this is merge.

File: test/CMakeLists.txt
Comment (Klaim) [reply test/CMakeLists.txt:82]: No because `/Wall` activates warning from dependencies, which is what the Turner points. It's not something we can easilly turn off without dependency-specific flags. That's why it's not recommended with msvc. It's not like gcc/clang's `-Wall` which does *not* activate all warnings, but only recomended ones.

**Review by JohanMabille — APPROVED** [review]

File: test/CMakeLists.txt
Comment (Klaim) [reply test/CMakeLists.txt:82]: It is not enough to disable the dependencies included through `"` :/

**Pull request merged**
